From 54fbece6b2beafabf1433e6d4e737828dfb7b700 Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Sun, 20 Mar 2016 18:42:03 +0200
Subject: [PATCH 282/538] ap806: Coresight: add DT bindings for Coresight
 components

AP806-A0 contains a trace subsystem that can be configured via Linux
drivers. This patch adds the DT bindings that describes the different
Coresight components in the trace subsystem and their connectivity.

Each CPU has an Embedded Trace Macrocell (ETM) which captures the trace
from CPU execution. This trace is passed to a per-cpu Embedded Trace
FIFO (ETF) of 2KB. All 4 ETFs are aggregated to a common 32KB ETF via a
Coresight funnel. Finally the ETF trace can be written to memory via an
Embbeded Trace Router (ETR).

Change-Id: I9b11884fd6b74dcbc336a551ce9bf98207d61f71
Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28370
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi | 267 ++++++++++++++++++++-
 1 file changed, 263 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi b/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
index 425cdd9..c293f13 100644
--- a/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
+++ b/arch/arm64/boot/dts/marvell/armada-ap806-quad.dtsi
@@ -54,36 +54,295 @@
 		#address-cells = <1>;
 		#size-cells = <0>;
 
-		cpu@000 {
+		cpu0: cpu@000 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x000>;
 			enable-method = "psci";
 		};
-		cpu@001 {
+		cpu1: cpu@001 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x001>;
 			enable-method = "psci";
 		};
-		cpu@100 {
+		cpu2: cpu@100 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x100>;
 			enable-method = "psci";
 		};
-		cpu@101 {
+		cpu3: cpu@101 {
 			device_type = "cpu";
 			compatible = "arm,cortex-a72", "arm,armv8";
 			reg = <0x101>;
 			enable-method = "psci";
 		};
 	};
+
 	ap806 {
 		config-space {
 			sdhci@d8000 {
 				status = "okay";
 			};
+			etm0@c40000 {
+				compatible = "arm,coresight-etm4x", "arm,primecell";
+				reg = <0xc40000 0x1000>;
+
+				cpu = <&cpu0>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				port {
+					etm0_out: endpoint {
+						remote-endpoint = <&etf0_in>;
+					};
+				};
+			};
+
+			etm1@d40000 {
+				compatible = "arm,coresight-etm4x", "arm,primecell";
+				reg = <0xd40000 0x1000>;
+
+				cpu = <&cpu1>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				port {
+					etm1_out: endpoint {
+						remote-endpoint = <&etf1_in>;
+					};
+				};
+			};
+
+			etm2@e40000 {
+				compatible = "arm,coresight-etm4x", "arm,primecell";
+				reg = <0xe40000 0x1000>;
+
+				cpu = <&cpu2>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				port {
+					etm2_out: endpoint {
+						remote-endpoint = <&etf2_in>;
+					};
+				};
+			};
+
+			etm3@f40000 {
+				compatible = "arm,coresight-etm4x", "arm,primecell";
+				reg = <0xf40000 0x1000>;
+
+				cpu = <&cpu3>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+				port {
+					etm3_out: endpoint {
+						remote-endpoint = <&etf3_in>;
+					};
+				};
+			};
+
+			etf0@809000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x809000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					port@0 {
+						reg = <0>;
+						etf0_in: endpoint {
+							slave-mode;
+							remote-endpoint = <&etm0_out>;
+						};
+					};
+
+					port@1 {
+						reg = <0>;
+						etf0_out: endpoint {
+							remote-endpoint = <&funnel_in_0>;
+						};
+					};
+				};
+			};
+
+			etf1@80a000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x80a000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					port@0 {
+						reg = <0>;
+						etf1_in: endpoint {
+							slave-mode;
+							remote-endpoint = <&etm1_out>;
+						};
+					};
+
+					port@1 {
+						reg = <0>;
+						etf1_out: endpoint {
+							remote-endpoint = <&funnel_in_1>;
+						};
+					};
+				};
+			};
+
+			etf2@80b000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x80b000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					port@0 {
+						reg = <0>;
+						etf2_in: endpoint {
+							slave-mode;
+							remote-endpoint = <&etm2_out>;
+						};
+					};
+
+					port@1 {
+						reg = <0>;
+						etf2_out: endpoint {
+							remote-endpoint = <&funnel_in_2>;
+						};
+					};
+				};
+			};
+
+			etf3@80c000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x80c000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					port@0 {
+						reg = <0>;
+						etf3_in: endpoint {
+							slave-mode;
+							remote-endpoint = <&etm3_out>;
+						};
+					};
+
+					port@1 {
+						reg = <0>;
+						etf3_out: endpoint {
+							remote-endpoint = <&funnel_in_3>;
+						};
+					};
+				};
+			};
+
+			funnel@809000 {
+				compatible = "arm,coresight-funnel", "arm,primecell";
+				reg = <0x809000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					/* funnel output port */
+					port@0 {
+						reg = <0>;
+						funnel_out: endpoint {
+							remote-endpoint = <&etf5_in>;
+						};
+					};
+
+					/* funnel input ports */
+					port@1 {
+						reg = <0>;
+						funnel_in_0: endpoint {
+							slave-mode;
+							remote-endpoint = <&etf0_out>;
+						};
+					};
+
+					port@2 {
+						reg = <1>;
+						funnel_in_1: endpoint {
+							slave-mode;
+							remote-endpoint = <&etf1_out>;
+						};
+					};
+
+					port@3 {
+						reg = <2>;
+						funnel_in_2: endpoint {
+							slave-mode;
+							remote-endpoint = <&etf2_out>;
+						};
+					};
+
+					port@4 {
+						reg = <3>;
+						funnel_in_3: endpoint {
+							slave-mode;
+							remote-endpoint = <&etf3_out>;
+						};
+					};
+				};
+			};
+
+			etf5@811000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x811000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+
+					port@0 {
+						reg = <0>;
+						etf5_in: endpoint {
+							slave-mode;
+							remote-endpoint = <&funnel_out>;
+						};
+					};
+
+					port@1 {
+						reg = <0>;
+						etf5_out: endpoint {
+							remote-endpoint = <&etr_in>;
+						};
+					};
+				};
+			};
+
+
+			etr@812000 {
+				compatible = "arm,coresight-tmc", "arm,primecell";
+				reg = <0x812000 0x1000>;
+				clocks = <&syscon 3>;
+				clock-names = "apb_pclk";
+
+				port {
+					etr_in: endpoint {
+						slave-mode;
+						remote-endpoint = <&etf5_out>;
+					};
+				};
+			};
 		};
 	};
 };
-- 
1.9.1

